modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java

```diff
@@ -467,77 +472,170 @@ private double[][] getAreaWidths(List<Node>managed, double height, boolean minim
-                    temp[i] = computeChildMaxAreaWidth(child, getMinBaselineComplement(), getMargin(child), height, shouldFillHeight);
-                    adjustingNumber++;
+                    maxWidths[i] = computeChildMaxAreaWidth(
+                        child, getMinBaselineComplement(), getMargin(child), height, shouldFillHeight);
+                    ++adjustingNumber;
                 } else {
-                    temp[i] = -1;
+                    maxWidths[i] = -1;
                 }
             }
+
+            // Adjust the children that are eligible in this phase and return early if the children
+            // fit the target width (so no second phase is required).
+            if (adjustWidthsWithinLimits(managed, currentWidths, maxWidths, targetWidth, adjustingNumber)) {
+                return;
+            }
         }
+    }
 
-        double available = extraWidth; // will be negative in shrinking case
-        outer:while (Math.abs(available) > 1 && adjustingNumber > 0) {
-            final double portion = snapPortionX(available / adjustingNumber); // negative in shrinking case
-            for (int i = 0, size = managed.size(); i < size; i++) {
-                if (temp[i] == -1) {
+    /**
+     * Resizes the children widths to fit the target width, while taking into account the resize limits
+     * for each child (their minimum and maximum width). This method will be called once when shrinking,
+     * and may be called twice when growing.
+     *
+     * @param managed the managed children
+     * @param currentWidths the current children widths
+     * @param limitWidths the max or min widths for each child, depending on whether we are growing or shrinking;
+     *                    a value of -1 means the child cannot be resized
+     * @param targetWidth sum of child widths and spacing
+     * @param adjustingNumber a number that indicates how many children can be resized
+     * @return {@code true} if the child widths were successfully resized to fit the target width;
+     *         {@code false} otherwise
+     */
+    private boolean adjustWidthsWithinLimits(
+            List<Node> managed, double[] currentWidths, double[] limitWidths, double targetWidth, int adjustingNumber) {
+        double totalSpacing = (managed.size() - 1) * snapSpaceX(getSpacing());
+
+        // Current total width and current delta are two important numbers that we continuously
+        // update as this method converges towards a solution.
+        double currentTotalWidth = snapSpaceX(sum(currentWidths, managed.size())) + totalSpacing;
+        double currentDelta = targetWidth - currentTotalWidth;
+
+        // We repeatedly apply the following algorithm as long as we have space left to
+        // distribute (currentDelta), as well as children that are eligible to grow or
+        // shrink (adjustingNumber).
+        while ((currentDelta > Double.MIN_VALUE || currentDelta < -Double.MIN_VALUE) && adjustingNumber > 0) {
```

Comment:
is there a guarantee that this while loop ends?

Reply:
Yes, in line 612 we return from this method once we can't resize any further.